ofrak_core/ofrak/gui/script_builder.py

```diff
@@ -0,0 +1,348 @@
+        Returns a subset of the most up-to-date version of the script for the session to which the
+        resource belongs, including only those actions whose type matches `target_type` or is UNDEF.
+        (UNDEF actions are actions such as variable assignment.)
+
+        Allows a script to be created that reproduces specific actions which can then be applied to
+        different files, such as all unpack steps to arrive at a particular resource.
+
+        :param target_type: ActionType to include in the subset script.
+
+        :return: List of strings where each entry is a line in the script.
+        """
+        root_resource = await self._get_root_resource(resource)
+        return self._get_script(root_resource.get_id(), target_type)
+
+    async def _get_root_resource(self, resource: Resource) -> Resource:
+        """
+        Maps a given resource to its root for efficient retrieval of the root resource because
+        getting the root resource is likely the most performed operation in ScriptBuilder.
+        """
+        resource_id = resource.get_id()
+        if resource_id in self.root_cache:
+            return self.root_cache[resource_id]
+        while len(list(await resource.get_ancestors())) != 0:
+            resource = await resource.get_parent()
+        self.root_cache[resource_id] = resource
+        return resource
+
+    async def _get_variable_from_session(self, resource: Resource) -> str:
+        root_resource = await self._get_root_resource(resource)
+        return self.script_sessions[root_resource.get_id()].resource_variable_names[
+            resource.get_id()
+        ]
+
+    async def _var_exists(self, resource: Resource) -> bool:
+        root_resource = await self._get_root_resource(resource)
+        session = self._get_session(root_resource.get_id())
+        return resource.get_id() in session.resource_variable_names
+
+    async def _add_action_to_session(self, resource, action, action_type) -> None:
+        root_resource = await self._get_root_resource(resource)
+        session = self._get_session(root_resource.get_id())
+        session.actions.append(ScriptAction(action_type, action))
+
+    async def _add_variable_to_session(self, resource: Resource, var_name: str) -> None:
+        root_resource = await self._get_root_resource(resource)
+        session = self._get_session(root_resource.get_id())
+        session.resource_variable_names[resource.get_id()] = var_name
+
+    async def _get_selector(self, resource: Resource) -> str:
+        root_resource = await self._get_root_resource(resource)
+        for ancestor in await resource.get_ancestors():
+            if (
+                ancestor.get_id()
+                in self.script_sessions[root_resource.get_id()].resource_variable_names
+            ):
+                break
+        attribute, attribute_value = await self._get_selectable_attribute(resource)
+        try:
+            result = await ancestor.get_only_descendant(
```

Comment:
I'm a little confused why the code is searching up through the ancestors here - the code chunk returned by the function only ever contains a `get_only_child` call, and my memory is that that is the intent - we only  select resources one "level" at a time. What is the purpose of this loop?

Reply:
I think the issue is that the code chunk returned needs to be updated from `get_only_child` to `get_only_descendant`. @dannyp303 please correct me if I'm wrong, but I think that this loop uses the nearest ancestor which has been previously uniquely ID'd as the starting point from which to uniquely ID the current resource, allowing us to select resources at any level (i.e., enabling `unpack_recursively`).

Reply:
@EdwardLarson resolving based on changes made to refactor this method.